crates/turbo-tasks/src/vc/cast.rs

```diff
@@ -27,19 +27,7 @@ where
     type Output = ReadRef<T>;
 
     fn cast(content: TypedCellContent) -> Result<Self::Output> {
-        Ok(
-            // Safety: the `VcValueType` implementor must guarantee that both `T` and
-            // `Repr` are #[repr(transparent)].
-            unsafe {
-                // Downcast the cell content to the expected representation type, then
-                // transmute it to the expected type.
-                // See https://users.rust-lang.org/t/transmute-doesnt-work-on-generic-types/87272/9
-                std::mem::transmute_copy::<
-                    ManuallyDrop<ReadRef<<T::Read as VcRead<T>>::Repr>>,
-                    Self::Output,
-                >(&ManuallyDrop::new(content.cast()?))
-            },
-        )
```

Comment:
It turns out that this transmute wasn't entirely correct. You can't do a straight transmute between two `ReadRef<T>`s because it contains a `triomphe::Arc<T>`, and the vtable half of `triomphe::Arc<T>`'s fat pointer is different depending on the type of `T`.

You need to use something like `unchecked_sidecast_triomphe_arc` so that the fat pointer is constructed correctly. I modified the `TypedCellContent::cast` method in this PR to do the cast of `Arc` for us before constructing the `ReadRef`.